freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClientExceptionUtil.java

```diff
@@ -1,35 +1,243 @@
+            ErrorCodes.GENERIC_ERROR
+    );
 
-    public static final int NOT_FOUND_ERROR_CODE = 4001;
+    private static Set<Integer> retryableHttpCodes = Set.of(
+            HttpStatus.UNAUTHORIZED,
+            HttpStatus.PROXY_AUTHENTICATION_REQUIRED,
+            HttpStatus.REQUEST_TIMEOUT,
+            HttpStatus.TOO_MANY_REQUESTS,
+            HttpStatus.INTERNAL_SERVER_ERROR,
+            HttpStatus.BAD_GATEWAY,
+            HttpStatus.SERVICE_UNAVAILABLE,
+            HttpStatus.GATEWAY_TIMEOUT,
+            HttpStatus.BANDWIDTH_LIMIT_EXCEEDED,
+            HttpStatus.NETWORK_AUTHENTICATION_REQUIRED
+    )
+            .stream()
+            .map(s -> s.value())
+            .collect(Collectors.toSet());
 
-    public static final int DUPLICATE_ENTRY_ERROR_CODE = 4002;
+    static {
+        for (ErrorCodes errorCode : ErrorCodes.values()) {
+            ERROR_CODES_LOOKUP.put(errorCode.value, errorCode);
+        }
+    }
 
     private FreeIpaClientExceptionUtil() {
     }
 
     public static boolean isNotFoundException(FreeIpaClientException e) {
-        return isExceptionWithErrorCode(e, Set.of(NOT_FOUND_ERROR_CODE));
+        return isExceptionWithErrorCode(e, Set.of(ErrorCodes.NOT_FOUND));
     }
 
     public static boolean isDuplicateEntryException(FreeIpaClientException e) {
-        return isExceptionWithErrorCode(e, Set.of(DUPLICATE_ENTRY_ERROR_CODE));
+        return isExceptionWithErrorCode(e, Set.of(ErrorCodes.DUPLICATE_ENTRY));
     }
 
-    public static boolean isExceptionWithErrorCode(FreeIpaClientException e, Set<Integer> errorCodes) {
-        return Optional.ofNullable(e.getCause())
+    public static boolean isExceptionWithErrorCode(FreeIpaClientException e, Set<ErrorCodes> errorCodes) {
+        return Optional.ofNullable(getAncestorCauseBeforeFreeIpaClientExceptions(e))
                 .filter(JsonRpcClientException.class::isInstance)
                 .map(JsonRpcClientException.class::cast)
                 .map(JsonRpcClientException::getCode)
+                .flatMap(c -> Optional.ofNullable(ERROR_CODES_LOOKUP.get(c)))
                 .filter(c -> errorCodes.contains(c))
                 .isPresent();
     }
+
+    public static FreeIpaClientException convertToRetryableIfNeeded(FreeIpaClientException e) {
+        if (isRetryable(e)) {
+            return new RetryableFreeIpaClientException(e.getLocalizedMessage(), e, e.getStatusCode());
+        } else {
+            return e;
+        }
+    }
+
+    private static boolean isRetryable(FreeIpaClientException e) {
+        return isExceptionWithErrorCode(e, RETRYABLE_HTTP_CODES) ||
+                (e.getStatusCode().isPresent() &&
+                retryableHttpCodes.contains(e.getStatusCode().getAsInt()));
+    }
+
+    public static Throwable getAncestorCauseBeforeFreeIpaClientExceptions(FreeIpaClientException e) {
+        while (e.getCause() instanceof FreeIpaClientException) {
+            e = (FreeIpaClientException) e.getCause();
```

Comment:
assigning new value to the method parameter seems bad practice to me. could you introduce new variable here?